lib/scholar/cluster/affinity_propagation.ex

```diff
@@ -0,0 +1,382 @@
+          ]
+        ),
+        num_clusters: Nx.tensor(2, type: :u64),
+        similarity_matrix: Nx.tensor(
+          [
+            [-971.5, -6162.0, -5358.0, -5499.0],
+            [-6162.0, -971.5, -1030.0, -913.0],
+            [-5358.0, -1030.0, -971.5, -31.0],
+            [-5499.0, -913.0, -31.0, -971.5]
+          ]
+        )
+      }
+  """
+  def prune(
+        %__MODULE__{
+          cluster_centers_indices: cluster_centers_indices,
+          cluster_centers: cluster_centers,
+          num_clusters: k,
+          similarity_matrix: similarity_matrix
+        } = model
+      ) do
+    k = Nx.to_number(k)
+
+    cluster_centers_indices =
+      Nx.to_flat_list(cluster_centers_indices) |> Enum.filter(&(&1 != -1)) |> Nx.tensor()
+
+    cluster_centers = Nx.take(cluster_centers, cluster_centers_indices)
+
+    labels = Nx.argmax(Nx.take(similarity_matrix, cluster_centers_indices, axis: 1), axis: 1)
+    labels = Nx.indexed_put(labels, Nx.new_axis(cluster_centers_indices, -1), Nx.iota({k}))
+
+    %__MODULE__{
+      model
+      | cluster_centers_indices: cluster_centers_indices,
+        cluster_centers: cluster_centers,
+        labels: labels
+    }
+  end
+
+  @doc """
+  Predict the closest cluster each sample in `x` belongs to.
+
+  ## Examples
+
+      iex> seed = 42
+      iex> x = Nx.tensor([[12,5,78,2], [1,5,7,32], [1,3,6,1], [1,2,5,2]])
+      iex> model = Scholar.Cluster.AffinityPropagation.fit(x, seed: seed)
+      iex> model = Scholar.Cluster.AffinityPropagation.prune(model)
+      iex> Scholar.Cluster.AffinityPropagation.predict(model, Nx.tensor([[1,6,2,6], [8,3,8,2]]))
+      #Nx.Tensor<
+        s64[2]
+        [1, 1]
+      >
+  """
+  defn predict(%__MODULE__{cluster_centers: cluster_centers} = _model, x) do
+    {num_clusters, num_features} = Nx.shape(cluster_centers)
+    {num_samples, _} = Nx.shape(x)
+    broadcast_shape = {num_samples, num_clusters, num_features}
+
+    Scholar.Metrics.Distance.euclidean(
+      Nx.new_axis(x, 1) |> Nx.broadcast(broadcast_shape),
+      Nx.new_axis(cluster_centers, 0) |> Nx.broadcast(broadcast_shape),
+      axes: [-1]
+    )
+    |> Nx.argmin(axis: 1)
+  end
+
+  defnp initialize_matrices(data, opts \\ []) do
+    {n, _} = Nx.shape(data)
+    self_preference = opts[:self_preference]
+
+    {similarity_matrix, affinity_matrix} =
+      initialize_similarities(data, self_preference: self_preference)
+
+    zero = Nx.tensor(0, type: Nx.type(similarity_matrix))
+    availability_matrix = Nx.broadcast(zero, {n, n})
+    responsibility_matrix = Nx.broadcast(zero, {n, n})
+
+    {availability_matrix, responsibility_matrix, similarity_matrix, affinity_matrix}
+  end
+
+  defn initialize_similarities(data, opts \\ []) do
+    {n, dims} = Nx.shape(data)
+    self_preference = opts[:self_preference]
+    t1 = Nx.reshape(data, {1, n, dims})
+    t2 = Nx.reshape(data, {n, 1, dims})
+
+    dist =
+      (t1 - t2)
+      |> Nx.pow(2)
+      |> Nx.sum(axes: [2])
```

Comment:
isn't this Scholar.Metrics.Distance.euclidean_squared?

Reply:
If I broadcast t1 and t2 correctly beforehand, then yes